hswidget/cof_widget.go

```diff
@@ -0,0 +1,668 @@
+	directionList.Size(indicatorSize).OnChange(p.onUpdate)
+
+	frameStrings := make([]string, 0)
+	for idx := range p.cof.Priority[state.viewerState.directionIndex] {
+		frameStrings = append(frameStrings, fmt.Sprintf("%d", idx))
+	}
+
+	frameString := frameStrings[state.viewerState.frameIndex]
+	frameList := giu.Combo("##"+p.id+"frame", frameString, frameStrings, &state.frameIndex)
+	frameList.Size(indicatorSize).OnChange(p.onUpdate)
+
+	return giu.Layout{
+		giu.TabBar("COFViewerTabs").Layout(giu.Layout{
+			giu.TabItem("Animation").Layout(p.makeAnimationTab()),
+			giu.TabItem("Layer").Layout(p.makeLayerTab(state, layerList)),
+			giu.TabItem("Priority").Layout(p.makePriorityTab(state, directionList, frameList)),
+		}),
+	}
+}
+
+// this should also probably be a method of COF
+func calculateDuration(cof *d2cof.COF) float64 {
+	const (
+		milliseconds = 1000
+	)
+
+	frameDelay := milliseconds / speedToFPS(cof.Speed)
+
+	return float64(cof.FramesPerDirection) * frameDelay
+}
+
+func (p *COFWidget) makeAnimationTab() giu.Layout {
+	const (
+		fmtFPS        = "FPS: %.1f"
+		fmtDuration   = "Duration: %.2fms"
+		fmtDirections = "Directions: %v"
+		strSpeed      = "Speed: "
+		maxSpeed      = 256
+	)
+
+	numDirs := p.cof.NumberOfDirections
+	fps := speedToFPS(p.cof.Speed)
+	duration := calculateDuration(p.cof)
+
+	strLabelDirections := fmt.Sprintf(fmtDirections, numDirs)
+	strLabelFPS := fmt.Sprintf(fmtFPS, fps)
+	strLabelDuration := fmt.Sprintf(fmtDuration, duration)
+
+	speed32 := int32(p.cof.Speed)
+	setSpeed := func() {
+		p.cof.Speed = int(speed32)
+
+		if speed32 >= maxSpeed {
+			p.cof.Speed = maxSpeed
+		}
+	}
+
+	speedLabel := giu.Label(strSpeed)
+	speedInput := giu.InputInt("##"+p.id+"CovViewerSpeedValue", &speed32).Size(speedInputW).OnChange(setSpeed)
+
+	return giu.Layout{
+		giu.Label(strLabelDirections),
+		p.layoutAnimFrames(),
+		giu.Line(speedLabel, speedInput),
+		giu.Label(strLabelFPS),
+		giu.Label(strLabelDuration),
+	}
+}
+
+func (p *COFWidget) makeLayerTab(state *COFState, layerList giu.Widget) giu.Layout {
+	addLayerButtonID := fmt.Sprintf("Add a new layer...##%sAddLayer", p.id)
+	addLayerButton := giu.Button(addLayerButtonID).Size(actionButtonW, actionButtonH)
+	addLayerButton.OnClick(func() {
+		p.CreateNewLayer()
+	})
+
+	deleteLayerButtonID := fmt.Sprintf("Delete current layer...##%sDeleteLayer", p.id)
+	deleteLayerButton := giu.Button(deleteLayerButtonID).Size(actionButtonW, actionButtonH)
+	deleteLayerButton.OnClick(func() {
+		const (
+			strPrompt  = "Do you really want to remove this layer?"
+			strMessage = "If you'll click YES, all data from this layer will be lost. Continue?"
+		)
+
+		fnYes := func() {
+			p.deleteCurrentLayer(state.viewerState.layerIndex)
+			state.mode = cofEditorModeViewer
+		}
+
+		fnNo := func() {
+			state.mode = cofEditorModeViewer
+		}
+
+		id := fmt.Sprintf("##%sDeleteLayerConfirm", p.id)
+		state.viewerState.confirmDialog = NewPopUpConfirmDialog(id, strPrompt, strMessage, fnYes, fnNo)
```

Comment:
we can replace with PopUpConfirmDialog literal

Reply:
Then why have the function `NewPopUpConfirmDialog` ? If we're not going to use it, it needs to be removed, and that can happen in another PR.